operators/src/processing/raster_kernel/tile_sub_query.rs

```diff
@@ -0,0 +1,390 @@
+            input_tile,
+            pool,
+            kernel_fn,
+        }
+    }
+}
+
+#[async_trait]
+impl<P, F> FoldTileAccu for RasterKernelAccu<P, F>
+where
+    P: Pixel,
+    f64: AsPrimitive<P>,
+    F: KernelFunction<P> + 'static,
+{
+    type RasterType = P;
+
+    /// now that we collected all the input tile pixels we perform the actual raster kernel
+    async fn into_tile(self) -> Result<RasterTile2D<Self::RasterType>> {
+        let kernel_fn = self.kernel_fn.clone();
+        let output_tile = crate::util::spawn_blocking_with_thread_pool(self.pool, move || {
+            apply_kernel_for_each_inner_pixel(&self.input_tile, &self.output_info, &kernel_fn)
+        })
+        .await?;
+
+        Ok(output_tile)
+    }
+
+    fn thread_pool(&self) -> &Arc<ThreadPool> {
+        &self.pool
+    }
+}
+
+/// Apply kernel function to all pixels of the inner input tile in the 9x9 grid
+fn apply_kernel_for_each_inner_pixel<P, F>(
+    input: &RasterTile2D<P>,
+    info_out: &TileInformation,
+    kernel_fn: &F,
+) -> RasterTile2D<P>
+where
+    P: Pixel,
+    f64: AsPrimitive<P>,
+    F: KernelFunction<P>,
+{
+    if input.is_empty() {
+        return RasterTile2D::new_with_tile_info(
+            input.time,
+            *info_out,
+            EmptyGrid::new(info_out.tile_size_in_pixels).into(),
+        );
+    }
+
+    let map_fn = |gidx: GridIdx2D| {
+        let GridIdx([y, x]) = gidx;
+
+        // TODO: better to feed kernel values one-by-one?
+        let mut kernel_matrix =
+            Vec::<Option<P>>::with_capacity(kernel_fn.kernel_size().number_of_elements());
+
+        let y_stop = y + kernel_fn.y_width() as isize;
+        let x_stop = x + kernel_fn.x_width() as isize;
+        // copy row-by-row all pixels in x direction into kernel matrix
+        for y_index in y..y_stop {
+            for x_index in x..x_stop {
+                kernel_matrix.push(input.get_at_grid_index_unchecked([y_index, x_index]));
+            }
+        }
+
+        kernel_fn.apply(&kernel_matrix)
+    };
+
+    // TODO: this will check for empty tiles. Change to MaskedGrid::from(…) to avoid this.
+    let out_data = GridOrEmpty::from_index_fn_parallel(&info_out.tile_size_in_pixels, map_fn);
+
+    RasterTile2D::new(
+        input.time,
+        info_out.global_tile_position,
+        info_out.global_geo_transform,
+        out_data,
+    )
+}
+
+fn create_enlarged_tile<P: Pixel, F: KernelFunction<P>>(
+    tile_info: TileInformation,
+    query_rect: RasterQueryRectangle,
+    pool: Arc<ThreadPool>,
+    tiling_specification: TilingSpecification,
+    kernel_fn: F,
+) -> RasterKernelAccu<P, F>
+where
+    f64: AsPrimitive<P>,
+{
+    // create an accumulator as a single tile that fits all the input tiles + some margin for the kernel size
+
+    let tiling = tiling_specification.strategy(
+        query_rect.spatial_resolution.x,
+        -query_rect.spatial_resolution.y,
+    );
+
+    // let origin_coordinate = tiling
```

Comment:
remove

Reply:
[fe84d11](https://github.com/geo-engine/geoengine/pull/644/commits/fe84d112615b6f6e4dd8a3bae110fa84a93a94f2) and [b6b3237](https://github.com/geo-engine/geoengine/pull/644/commits/b6b323705d0d990d2ce33577c0f17192939cefcb)